Allow for remote names other than `origin`.

RomeoV opened the pull request:
We introduce `getremotename` that tries to check the `remote.pushDefault` option to determine the remote name, and falls back to `origin` if it's not set.

Fixes #2464 and perhaps #1916, as well as https://discourse.julialang.org/t/creating-documentation-when-remote-name-is-not-origin/110466.

mortenpi commented on src/utilities/utilities.jl:
```diff
@@ -421,7 +421,31 @@ const julia_remote = Remotes.GitHub("JuliaLang", "julia")
 Stores the memoized results of [`getremote`](@ref).
 """
-const GIT_REMOTE_CACHE = Dict{String,Union{Remotes.Remote,Nothing}}()
+const GIT_REMOTENAME_CACHE = Dict{String,Union{String,Nothing}}()
```
I don't think we expect `nothing` values here?

```suggestion
const GIT_REMOTENAME_CACHE = Dict{String,String}()
```

mortenpi commented:
Generally, I think trying to use `pushDefault` makes sense. But I have a couple of concerns:

1. It will warn when `remote.pushDefault` is not set. However, it doesn't seem that `remote.pushDefault` is set in the normal case (e.g. you have just `git clone`-d a repo). So it seems that 99% of the time this code will warn (you can see that in the CI logs too), which is not great.

2. I also wonder what is the right behavior when you do have `origin`, but `pushDefault` points to a different remote? I generally don't set `pushDefault`, but the use case I can imagine is when I copy a repo that I don't have access to, then fork it on GitHub and set `pushDefault` to the fork under my user. But if I were to build docs locally, I think I would prefer if it would pick `origin` over my fork. So maybe it should use `origin` if available, and if there is not remote called `origin`, it will pick the second best on based on `pushDefault`.

   Or maybe first step could be to just pick the only remote that is available, even if it is not called `origin`? And only use `pushDefault` if you have multiple remotes, none are called `origin`, and then use `pushDefault` to disambiguate?

3. It would be good to add a note somewhere on the [`docs/src/lib/remote-links.md`](https://github.com/JuliaDocs/Documenter.jl/blob/master/docs/src/lib/remote-links.md) manual page about this behavior.
4. Ideally there would also be a test. We have some related tests in [`test/repolinks.jl`](https://github.com/JuliaDocs/Documenter.jl/blob/master/test/repolinks.jl)

RomeoV commented:
Thanks for the thoughts. I have changed the workflow to default to `origin`, then `remotes.pushDefault` with a warning, and then just `""` (which was the behavior before) with another warning.
I also added a related comment to the docs, and wrote some tests.

fingolfin commented on src/utilities/utilities.jl:
```diff
@@ -421,7 +421,45 @@ const julia_remote = Remotes.GitHub("JuliaLang", "julia")
+"""
+$(TYPEDSIGNATURES)
+
+Determines the GitHub remote of a directory, returninf 'origin' if that remote  exists,
```
```suggestion
Determines the GitHub remote of a directory, returning `origin` if that remote  exists,
```

fingolfin commented on src/utilities/utilities.jl:
```diff
@@ -421,7 +421,45 @@ const julia_remote = Remotes.GitHub("JuliaLang", "julia")
+            @assert !isempty(pushdefault)
+            @warn "Remote 'origin' not found. Using `git config --get remote.pushDefault` instead, which is set to '$(pushdefault)'."
+            return pushdefault
+        catch ; end
```
I suspect this PR needs to be reformatted with `runic`

fingolfin commented on src/utilities/utilities.jl:
```diff
@@ -421,7 +421,45 @@ const julia_remote = Remotes.GitHub("JuliaLang", "julia")
+function getremotename(dir::AbstractString)
+    isdir(dir) || return nothing
+    return get!(GIT_REMOTENAME_CACHE, dir) do
+        try
+            originurl = readchomp(setenv(`$(git()) config --get remote.origin.url`; dir=dir))
+            @assert !isempty(originurl)
+            return "origin"
+        catch; end
```
You seem to be using `@assert` plus `try/catch` for control flow. That seems like a bad idea for multiple reasons (e.g. future Julia version might allow disabling `@assert` checks, which is, I think, mainly intended to enforce to check runtime invariants, which is not what this code does).

How about just e.g. this
```suggestion
        originurl = readchomp(setenv(`$(git()) config --get remote.origin.url`; dir=dir))
        if !isempty(originurl)
            return "origin"
        end
```

fingolfin replied:
Same applies to the rest of the code, of course

fingolfin commented on src/utilities/utilities.jl:
```diff
@@ -616,7 +655,7 @@ it out automatically.
 to construct the warning messages.
 """
-function git_remote_head_branch(varname, root; remotename = "origin", fallback = "master")
+function git_remote_head_branch(varname, root; fallback = "master")
```
The docstring still references `remotename`. So either that needs to be adjusted. Alternatively, keep the kwarg here, but with a new default value `nothing`; and then later only call `getremotename` if `remotename === nothing` ?

fingolfin commented:
@RomeoV are you still interested in this? If so, some conflicts need to be resolved now and a bunch of comments should be addressed.

If not, let's close it?
